// sqlite-wasm/lib/sqlite3-types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ReturnCode;
(function (ReturnCode) {
    ReturnCode[ReturnCode["OK"] = 0] = "OK";
    ReturnCode[ReturnCode["ERROR"] = 1] = "ERROR";
    ReturnCode[ReturnCode["INTERNAL"] = 2] = "INTERNAL";
    ReturnCode[ReturnCode["PERM"] = 3] = "PERM";
    ReturnCode[ReturnCode["ABORT"] = 4] = "ABORT";
    ReturnCode[ReturnCode["BUSY"] = 5] = "BUSY";
    ReturnCode[ReturnCode["LOCKED"] = 6] = "LOCKED";
    ReturnCode[ReturnCode["NOMEM"] = 7] = "NOMEM";
    ReturnCode[ReturnCode["READONLY"] = 8] = "READONLY";
    ReturnCode[ReturnCode["INTERRUPT"] = 9] = "INTERRUPT";
    ReturnCode[ReturnCode["IOERR"] = 10] = "IOERR";
    ReturnCode[ReturnCode["CORRUPT"] = 11] = "CORRUPT";
    ReturnCode[ReturnCode["NOTFOUND"] = 12] = "NOTFOUND";
    ReturnCode[ReturnCode["FULL"] = 13] = "FULL";
    ReturnCode[ReturnCode["CANTOPEN"] = 14] = "CANTOPEN";
    ReturnCode[ReturnCode["PROTOCOL"] = 15] = "PROTOCOL";
    ReturnCode[ReturnCode["EMPTY"] = 16] = "EMPTY";
    ReturnCode[ReturnCode["SCHEMA"] = 17] = "SCHEMA";
    ReturnCode[ReturnCode["TOOBIG"] = 18] = "TOOBIG";
    ReturnCode[ReturnCode["CONSTRAINT"] = 19] = "CONSTRAINT";
    ReturnCode[ReturnCode["MISMATCH"] = 20] = "MISMATCH";
    ReturnCode[ReturnCode["MISUSE"] = 21] = "MISUSE";
    ReturnCode[ReturnCode["NOLFS"] = 22] = "NOLFS";
    ReturnCode[ReturnCode["AUTH"] = 23] = "AUTH";
    ReturnCode[ReturnCode["FORMAT"] = 24] = "FORMAT";
    ReturnCode[ReturnCode["RANGE"] = 25] = "RANGE";
    ReturnCode[ReturnCode["NOTADB"] = 26] = "NOTADB";
    ReturnCode[ReturnCode["NOTICE"] = 27] = "NOTICE";
    ReturnCode[ReturnCode["WARNING"] = 28] = "WARNING";
    ReturnCode[ReturnCode["ROW"] = 100] = "ROW";
    ReturnCode[ReturnCode["DONE"] = 101] = "DONE";
    ReturnCode[ReturnCode["INTEGER"] = 1] = "INTEGER";
    ReturnCode[ReturnCode["FLOAT"] = 2] = "FLOAT";
    ReturnCode[ReturnCode["TEXT"] = 3] = "TEXT";
    ReturnCode[ReturnCode["BLOB"] = 4] = "BLOB";
    ReturnCode[ReturnCode["NULL"] = 5] = "NULL";
    ReturnCode[ReturnCode["UTF8"] = 1] = "UTF8";
})(ReturnCode = exports.ReturnCode || (exports.ReturnCode = {}));
